### PR TITLE
Optimize multithreading performance of RE2J

### PR DESCRIPTION
**Pull request by ghost**
Hi Alan, 

We have put up this pull request for multithreading performance fixes.
We noticed that the contention overhead for RE2.get() function that returns 
Machine objects from the Machine cache is  greater than 
the slowdown that may have been created by instantiating newer Machine objects. 

To illustrate: the test case when we noticed this was in a scenario with Presto (database 
that queries hive data) where 128 concurrent threads were using RE2 library at a time. Over the 
lifetime of a regular expression query, a single process calls pattern.matcher() and matcher.find() 
more than a million times. All these calls are made by multiple threads (128) at a time. 

We noticed that when removing the machine cache, time taken by a regular expression query 
dropped down to 10min from 40min when using the machine cache. We verified this 
with/without the machine cache and also used yourkit profiler. All types of regular 
expression queries performed faster. 

This change has no effect on programs that use single thread. Additionally, we have 
also put two small changes in this pull request which are mostly minor optimizations. 

To summarize, this pull request attempts to: 
1. remove machine cache to avoid contention when retrieving cached machine objects
2. use array of Inst objects in Prog to optimize faster access to compiled Insts
3. declare RE2 and threadpool as final in Machine.java

Let us know your comments. Rebecca had also briefly discussed these improvements here:  https://groups.google.com/forum/#!topic/re2j-discuss/3iEzpgxiI4k

### CODE REVIEW

**Comment by kkc00** [comment]
I have updated the pull request by adding a new commit that removes the change related to 
adding Inst array. We will revise it so that it is only created once in Prog object and all 
machine objects access it via Prog object. 

So, the only changes now are 1) removing machine cache, and 2) declaring RE2 and threadpool as final. 

Let us know your comments. If it looks good, then I will squash the two commits in the pull request to create a single commit. For now I created the separate commit to preserve the commit comments.

**Comment by kkc00** [comment]
Hi Alan, 

Just pinging you in case you haven't had a chance to go through the updated pull request. Let us know if you would be ok with merging this, I will then squash the commits into one. 

Also, we have other updates in pipeline especially regarding adding other matching engines (eg. DFA) from the original RE2. We can create separate issues to discuss them.

**Comment by alandonovan** [comment]
Hi Kamal,  I'm sad to say this but I am finding it very hard to find time
to review RE2/J changes.  Rebecca and Felix (cc'd) have kindly offered to
take on the maintenance of this package.  Can I suggest you direct your
query to them?

cheers (and thanks to Rebecca and Felix)
alan

**Comment by rschlussel** [comment]
I don't have write access, but this looks good.

**Comment by sjamesr** [comment]
@rschlussel hi Rebecca, I gave you collaborator access, you should be able to merge the change into master when all the checks pass

**Comment by rschlussel** [comment]
Thanks!

**Comment by kkc00** [comment]
Thanks a lot guys. I have squashed the two commits into one, with build passing, and we are in process of signing the CLA. Thanks again.

**Comment by rschlussel** [comment]
closing this because CLA isn't signed and the user is gone.
